openssl/error.go

```diff
@@ -0,0 +1,59 @@
+// Copyright (C) 2018 NuCypher
+//
+// This file is part of goUmbral.
+//
+// goUmbral is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// goUmbral is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with goUmbral. If not, see <https://www.gnu.org/licenses/>.
+package openssl
+
+// #include "shim.h"
+import "C"
+import (
+    "fmt"
+)
+
+const ERR_R_FATAL = 64
+
+type OpenSSLError struct {
+    code uint64
+    library string
+    function string
+    reason string
+    fatal bool
+}
+
+func (m *OpenSSLError) Error() string {
+    if m.fatal {
+        return fmt.Sprintf("OpenSSL FATAL Error: %x:%s:%s:%s",
+            m.code, m.library, m.function, m.reason)
+    } else {
+        return fmt.Sprintf("OpenSSL NON_FATAL Error: %x:%s:%s:%s",
```

Comment:
:heart_eyes:

Reply:
This is my favorite part of goUmbral haha. I loved making nice structs for the OpenSSL errors to hang out in.